src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessor.cs

```diff
@@ -36,24 +38,33 @@ protected AbstractQueriesHandlerProcessor([NotNull] TRequestHandler requestHandl
 
     protected abstract RavenConfiguration Configuration { get; }
 
-    protected RequestTimeTracker CreateRequestTimeTracker()
+    internal RequestTimeTracker CreateRequestTimeTracker()
     {
         return new RequestTimeTracker(HttpContext, Logger, NotificationCenter, Configuration, "Query");
     }
 
-    public async ValueTask<IndexQueryServerSide> GetIndexQueryAsync(JsonOperationContext context, HttpMethod method, RequestTimeTracker tracker, bool addSpatialProperties = false)
+    internal OperationCancelToken CreateHttpRequestBoundTimeLimitedOperationTokenForQuery() => RequestHandler.CreateHttpRequestBoundTimeLimitedOperationTokenForQuery();
+    
+    internal ValueTask<IndexQueryServerSide> GetIndexQueryAsync(JsonOperationContext context, HttpMethod queryMethod, RequestTimeTracker tracker, bool addSpatialProperties)
     {
-        if (method == HttpMethod.Get)
-            return await ReadIndexQueryAsync(context, tracker, addSpatialProperties);
-
-        var json = await context.ReadForMemoryAsync(_stream, "index/query");
-
+        if (QueryMethod == HttpMethod.Get)
+            return ValueTask.FromResult(ReadIndexQueryForGet(context, tracker, addSpatialProperties));
+        
+        return ReadIndexQueryForPost(context, tracker, addSpatialProperties);
+    }
+    
+    internal async ValueTask<IndexQueryServerSide> ReadIndexQueryForPost(JsonOperationContext context, RequestTimeTracker tracker, bool addSpatialProperties)
+    {
+        Debug.Assert(QueryMethod != HttpMethod.Get);
+        var readJsonTask = context.ReadForMemoryAsync(_stream, "index/query");
+        var json = readJsonTask.IsCompletedSuccessfully ? readJsonTask.Result : await readJsonTask;
```

Comment:
I'd double check the generated IL. As far as I remember this check is emitted already. The question is how likely is it that it's `.IsCompletedSuccessfully ` and weather you want to delegate the `await readJsonTask` to a separate async method